File: aquatrack_app/lib/features/missions/models/quest.dart
/// Quest as returned by GET /quests (derived progress + claim state).
class Quest {
  final String id;
  final String period; // "daily" | "weekly"
  final String name;
  final String description;
  final String unit;
  final int progress;
  final int target;
  final int rewardXp;
  final int rewardCoin;
  final bool isBonus;
  final bool isChest;
  final bool done;
  final bool claimed;

  const Quest({
    required this.id,
    required this.period,
    required this.name,
    required this.description,
    required this.unit,
    required this.progress,
    required this.target,
    required this.rewardXp,
    required this.rewardCoin,
    required this.isBonus,
    required this.isChest,
    required this.done,
    required this.claimed,
  });

  factory Quest.fromJson(Map<String, dynamic> json) => Quest(
        id: json['id'] as String,
        period: json['period'] as String,
        name: json['name'] as String,
        description: json['description'] as String,
        unit: (json['unit'] as String?) ?? '',
        progress: (json['progress'] as num).toInt(),
        target: (json['target'] as num).toInt(),
        rewardXp: (json['reward_xp'] as num).toInt(),
        rewardCoin: (json['reward_coin'] as num).toInt(),
        isBonus: json['is_bonus'] as bool? ?? false,
        isChest: json['is_chest'] as bool? ?? false,
        done: json['done'] as bool? ?? false,
        claimed: json['claimed'] as bool? ?? false,
      );
}

/// One day in the 7-day week strip.
class WeekDayStatus {
  final String dayLabel; // "T2" … "CN"
  final String dateIso; // "2026-05-25"
  final String status; // "done" | "partial" | "today" | "future"
  final int? progressPct; // 0–100, null for future

  const WeekDayStatus({
    required this.dayLabel,
    required this.dateIso,
    required this.status,
    this.progressPct,
  });

  factory WeekDayStatus.fromJson(Map<String, dynamic> json) => WeekDayStatus(
        dayLabel: json['day_label'] as String,
        dateIso: json['date_iso'] as String,
        status: json['status'] as String,
        progressPct: (json['progress_pct'] as num?)?.toInt(),
      );
}

/// Full payload of GET /quests: quests grouped by period + header balances.
class QuestsData {
  final List<Quest> daily;
  final List<Quest> weekly;
  final int coins;
  final int totalXp;
  final int currentLevel;
  final int currentStreak;
  final List<WeekDayStatus> weekStrip;
  final DateTime dailyResetAt;
  final DateTime weeklyResetAt;

  const QuestsData({
    required this.daily,
    required this.weekly,
    required this.coins,
    required this.totalXp,
    required this.currentLevel,
    required this.currentStreak,
    required this.weekStrip,
    required this.dailyResetAt,
    required this.weeklyResetAt,
  });

  factory QuestsData.fromJson(Map<String, dynamic> json) => QuestsData(
        daily: (json['daily'] as List)
            .map((e) => Quest.fromJson(e as Map<String, dynamic>))
            .toList(),
        weekly: (json['weekly'] as List)
            .map((e) => Quest.fromJson(e as Map<String, dynamic>))
            .toList(),
        coins: (json['coins'] as num?)?.toInt() ?? 0,
        totalXp: (json['total_xp'] as num?)?.toInt() ?? 0,
        currentLevel: (json['current_level'] as num?)?.toInt() ?? 1,
        currentStreak: (json['current_streak'] as num?)?.toInt() ?? 0,
        weekStrip: (json['week_strip'] as List? ?? [])
            .map((e) => WeekDayStatus.fromJson(e as Map<String, dynamic>))
            .toList(),
        dailyResetAt: DateTime.parse(json['daily_reset_at'] as String),
        weeklyResetAt: DateTime.parse(json['weekly_reset_at'] as String),
      );
}

File: aquatrack_backend/app/schemas/quest.py
from typing import List, Optional

from pydantic import BaseModel


class QuestOut(BaseModel):
    """A single quest with derived progress and claim state."""

    id: str
    period: str  # "daily" | "weekly"
    name: str
    description: str
    unit: str
    progress: int
    target: int
    reward_xp: int
    reward_coin: int
    is_bonus: bool
    is_chest: bool
    done: bool
    claimed: bool


class WeekDayStatus(BaseModel):
    """Status of one day in the current ISO week for the 7-day strip."""

    day_label: str  # "T2" … "CN"
    date_iso: str  # "2026-05-25"
    status: str  # "done" | "partial" | "today" | "future"
    progress_pct: Optional[int] = None  # 0–100, None for future days


class QuestsResponse(BaseModel):
    """All quests plus the user's reward balances for the header."""

    daily: List[QuestOut]
    weekly: List[QuestOut]
    coins: int
    total_xp: int
    current_level: int
    current_streak: int
    week_strip: List[WeekDayStatus]
    daily_reset_at: str  # ISO-8601 local datetime of next daily reset
    weekly_reset_at: str  # ISO-8601 local datetime of next weekly reset


class ClaimResponse(BaseModel):
    quest_id: str
    reward_xp: int
    reward_coin: int
    total_xp: int
    coins: int
    current_level: int
    # Level-Up Rewards (ADR 0008): present so a claim that crossed a level can
    # drive the celebration (carries coins_awarded for this claim).
    level_progress: Optional[dict] = None
